src/hotspot/share/opto/library_call.cpp

```diff
@@ -7776,8 +7776,22 @@ bool LibraryCallKit::inline_blackhole() {
   assert(callee()->is_empty(), "Should have been checked before: only empty methods here");
   assert(callee()->holder()->is_loaded(), "Should have been checked before: only methods for loaded classes here");
 
+  // Like GraphKit::insert_mem_bar, this code takes and produces control
+  // to pin the Blackhole in the graph. Unlike GraphKit::insert_mem_bar,
+  // this code only uses memory, but does not produce it, which avoids
+  // the unfortunate side effects like breaking the optimizations across
+  // the blackhole.
+
+  Node* mem = reset_memory();
+  set_all_memory(mem);
+
+  MemBarNode* mb = MemBarNode::make(C, Op_Blackhole);
+  mb->init_req(TypeFunc::Control, control());
+  mb->init_req(TypeFunc::Memory,  mem);
```

Comment:
That's a good question. I don't think it needs memory. I disconnected the input memory in new commit as well, and all tests seem fine. This also allows to simplify anti-dependence logic, as `Blackhole` does not have the asymmetry of "takes memory, but does not produce it" anymore.

AFAICS, `Blackhole` being a subclass of `MemBar` helps to avoid additional checks in other places in compiler, where we can test for `is_MemBar`. I can see if we can move `Blackhole` out of `MemBar` class without messing things up.

Reply:
Yeah, I also cannot see why it would need the memory input so long as it has the data dependency.

The key question is whether it's status as a membar is stopping things being re-ordered around it. Changing it's type might immediately show a problem ... or it might not (the problem with a 'suck it and see' approach is that you need to be sure you have sucked every type of sweetie in the sweet tin).

Reply:
I whipped up this patch that pulls `Blackhole` from `MemBarNode` to be the more generic `MultiNode`: https://cr.openjdk.java.net/~shade/8296545/blackhole-cfg-1.patch -- it seems to "work fine" on adhoc tests. But, I am still a bit uneasy to unhook blackhole from membar, on the off-chance it matters in some non-obvious way.

Reply:
> can probably even be Node, if I understand how should control-output-only nodes be defined

Yes, `MultiNode` is specifically for cases when a node produces multiple results. 

FTR I looked through the code base for `is_MemBar` usages and didn't spot any problematic cases, if you clear away memory input. But I'm fine with keeping it a membar.

Reply:
Yeah, I agree that your `blackhole-cfg-1.patch` is the lowest friction way to achieve the goal.

There are some pure control nodes (e.g., `Region`), but they are treated specially during code motion. So, special cases for `Blackhole` would be needed there.

Reply:
All right, I pushed the version that does `MultiNode` and passes all my tests.